File: throughput.py
#!/usr/bin/env python
import sys
import socket
import numpy as np
import math
import time
from time import sleep

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
args = sys.argv
mode = args[1]
# server side: ./throughput.py server
# cilent side: ./throughput.py <server's ip address>

size = 1
maxsize = 1024*1024
limit = math.log(maxsize,2)

if mode == "server": 
	s.bind((socket.gethostname(), 5000))
	s.listen(5)
	s.setblocking(1)
	print('listening...')
	conn, addr = s.accept()
	for _ in range(int(limit)+1):
		while True:
			rbuf = conn.recv(maxsize)
			if not rbuf: break	
		#print(size,'[B]')
		#size *= 2
	conn.close()
	s.close()
else:
	s.connect((mode, 5000))
	for _ in range(int(limit)+1):
		data = np.array([1 for x in range(size)], dtype=str)
		iter = 100000 if size < 1024*2 else(10000 if size <1024*32 else 100)
		start = time.time()
		for _ in range(iter):
			s.sendall(data)
		elasped_time = (time.time() - start)
		th = 8 * data.size * iter / elasped_time / 1024 / 1024 
		print(data.size,'[B],\t', elasped_time,'[s],\t', iter,'[loops],\t', th,'[Mbps],')
		size *= 2
	s.close()

File: throughput.go
package main

import (
	"fmt"
	"net"
	"log"
	"os"
	_"bufio"
	"io"
	_"io/ioutil"
	_"syscall"
	"time"
	_"bytes"
)

func main() {
	switch os.Args[1] {
	case "server":
		SocketServer()
	case "client":
		SocketClient()
	default:
		panic("bad command")
	}
}

func SocketServer() {
	maxsize := 1024*1024
	s, err := net.Listen("tcp4", ":5000")	
	if err != nil {
		log.Fatal(err)
	}
	defer s.Close()
	fmt.Printf("listening...\n")
	
	conn, err := s.Accept()
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close()

	//rbuf := make([]byte, maxsize)
	for size:=1; size<=maxsize; size*=2 {
		iter := 1
		if size < 1024*2 {
			iter = 100000
		} else if size < 1024*32 {
			iter = 10000
		} else {
			iter = 100
		}
		rbuf := make([]byte, size)
		for i:=1; i<=iter; i++ {
			_ = recv(rbuf, conn)
		}
		fmt.Printf("%v[B]\n", size)
	}
}

func SocketClient() {
	maxsize := 1024*1024
	ip := os.Args[2]
	conn, err := net.Dial("tcp4", ip + ":5000")
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close()

	for size:=1; size<=maxsize; size*=2 {
		iter := 1
		if size < 1024*2 {
			iter = 100000
		} else if size < 1024*32 {
			iter = 10000
		} else {
			iter = 100
		}
		sbuf := make([]byte, size)
		sbuf[0] = 1
		start := time.Now()
		for i:=1; i<=iter; i++ {
			send(sbuf, conn)
		}
		end := time.Now()
		elasped_time := end.Sub(start).Seconds()
		th := 8*float64(size)*float64(iter)/elasped_time/1024/1024
		fmt.Printf("%v[B]\t, %v[s]\t, %v[loops]\t, %v[Mbps]\n", size, elasped_time, iter, th)
	}	
}

// http://golang.jp/pkg/net1
//func recv(rbuf []byte, sbuf[]byte, conn io.Reader) (int){
//func recv(rbuf []byte, sbuf []byte, conn net.Conn, f *os.File) (int){
func recv(rbuf []byte, conn net.Conn) (int) {
	recvbyte := 0	
	for {
		//recvbyte = 0	
		_, err := io.ReadFull(conn, rbuf)
		//n, err := conn.Read(rbuf)
		//_, _ = f.Write(rbuf)
		//_, err = io.Copy(ioutil.Discard, conn)
		//recvbyte += n
		//fmt.Println(recvbyte)
		break
		if err != nil {
			if err == io.EOF {
				break
			} else {
				log.Fatal(err)
			}
		}
	}
	return recvbyte
}

//func send(sbuf []byte, rbuf []byte, conn io.Writer) (int){
//func send(sbuf []byte, rbuf []byte, conn net.Conn) (int){
func send(sbuf []byte, conn net.Conn) {
	_, err := conn.Write(sbuf)
	if err != nil {
		if err != nil {
			log.Fatal(err)
		}
	}
}
